### PR TITLE
[SPARK-38531][SQL] Fix the condition of "Prune unrequired child index" branch of ColumnPruning

### PR DESCRIPTION
**Pull request by minyyy**
### What changes were proposed in this pull request?

The "prune unrequired references" branch has the condition:

`case p @ Project(_, g: Generate) if p.references != g.outputSet => `

This is wrong as generators like Inline will always enter this branch as long as it does not use all the generator output.

Example:

input: <col1: array<struct<a: struct<a: int, b: int>, b: int>>>

Project(a.a as x)
\- Generate(Inline(col1), ..., a, b)

p.references is [a]
g.outputSet is [a, b]

This bug makes us never enter the GeneratorNestedColumnAliasing branch below thus miss some optimization opportunities. This PR changes the condition to check whether the child output is not used by the project and it is either not used by the generator or not already put into unrequiredChildOutput.

### Why are the changes needed?
The wrong condition prevents some expressions like Inline, PosExplode from being optimized by rules after it. Before the PR, the test query added in the PR is not optimized since the optimization rule is not able to apply to it. After the PR the optimization rule can be correctly applied to the query.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Unit tests.

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasingSuite.scala [hunk 1]
```diff
@@ -313,6 +313,25 @@ class NestedColumnAliasingSuite extends SchemaPruningTest {
     comparePlans(optimized, expected)
   }
 
+  test("Nested field pruning for Project and PosExplode") {
```
Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasingSuite.scala:316]: Can you add JIRA ticket as usual? Thanks.

Comment (viirya) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasingSuite.scala:316]: Oh, BTW, this rule is not for NestedColumnAliasing actually. Can you move it to `ColumnPruningSuite`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -831,8 +831,16 @@ object ColumnPruning extends Rule[LogicalPlan] {
...
-      val requiredAttrs = p.references -- g.producedAttributes ++ g.generator.references
+    // There are 2 types of pruning here:
+    // 1. For attributes in g.child.outputSet that is not used by the generator nor the project,
+    //    we directly remove it from the output list of g.child.
+    // 2. For attributes that is not used by the project but it is used by the generator, we put
+    //    it in g.unrequiredChildIndex to save memory usage.
+    case p @ Project(_, g: Generate) if g.child.output.zipWithIndex.exists(
+      pair =>
+      !p.references.contains(pair._1) &&
+        (!g.generator.references.contains(pair._1) || !g.unrequiredChildIndex.contains(pair._2))) =>
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:842]: Hmm, this is not for NestedColumnAliasing only but ColumnPruning. Is it possible to add a test in `ColumnPruningSuite` too?

Comment (minyyy) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:842]: Done. But the 2 tests are the same, if you want I can remove one test.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasingSuite.scala (hunk 1 above)
Comment (minyyy) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasingSuite.scala]: Done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 2 above)
Comment (sigmod) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:842]: Does it work if we switch the order case branches? Since the rule is a fix-point batch, it can hit `GeneratorNestedColumnAliasing` in the first iteration and Project(_, g: Generate) in the next?

```
    // prune unrequired nested fields from `Generate`.
    // prioritize this case branch over  Project(_, g: Generate) for SPARK-38531
    case GeneratorNestedColumnAliasing(rewrittenPlan) => rewrittenPlan
    
   // prune unrequired references
    case p @ Project(_, g: Generate) if p.references != g.outputSet => ...
```

Comment (minyyy) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:842]: Even if we switch the order, we still have to fix the condition otherwise this rule prevents following rules from being executed.

Comment (sigmod) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:842]: Some thoughts here:
- the condition sounds right to me;
- however, maybe it's just me, but it seems fragile for me to maintain two versions of the same logic forward - in `if` guard and in branch body;
- `unrequiredChildIndex` is a `Seq` rather than a set. Can it result in quadratic behaviors (probably in extreme cases where both `g.child.output` and `g.generator.references` are large)? 

Does it make sense to use an extractor pattern?

Comment (minyyy) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: I changed to the extractor pattern. What part did you refer to for the `quadratic behaviors`?

Comment (sigmod) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: > I changed to the extractor pattern. What part did you refer to for the quadratic behaviors?

`g.unrequiredChildIndex.contains(pair._2)` is o(n) where n is the size of g.unrequiredChildIndex, so that it will be o(m * n) where m is the g.child.output.size.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala [hunk 3]
```diff
@@ -312,6 +312,32 @@ object NestedColumnAliasing {
   }
 }
 
+object GeneratorUnrequiredChildrenPruning {
+  def unapply(plan: LogicalPlan): Option[LogicalPlan] = plan match {
+    case p@Project(_, g: Generate) =>
```
Comment (sigmod) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:317]: Nit: spaces around @ ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala [hunk 4]
```diff
@@ -312,6 +312,32 @@ object NestedColumnAliasing {
...
+      val unrequiredIndices = newChild.output.zipWithIndex.filter(t => unrequired.contains(t._1))
+        .map(_._2)
+      if (unrequiredIndices.toSet != g.unrequiredChildIndex.toSet) {
+        pruned = true
+      }
+      if (pruned) {
+        Some(p.copy(child = g.copy(child = newChild, unrequiredChildIndex = unrequiredIndices)))
+      } else {
+        None
+      }
```
Comment (sigmod) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:336]: There seems some logic change in the new code: https://www.diffchecker.com/wEi4OAYD, e.g., how `requiredAttrs` is obtained.

Is it possible to keep the change "mechanical", e.g.,

```
case p @ Project(_, g: Generate) if p.references != g.outputSet =>
val requiredAttrs = p.references -- g.producedAttributes ++ g.generator.references	
val newChild = prunedChild(g.child, requiredAttrs)	
val unrequired = g.generator.references -- p.references	
val unrequiredIndices = newChild.output.zipWithIndex.filter(t => unrequired.contains(t._1))	
        .map(_._2)	
val newP = if (newChild!=g.Child || unrequiredIndices.toSet != g.unrequiredChildIndex.toSet) {
    p.copy(child = g.copy(child = newChild, unrequiredChildIndex = unrequiredIndices))
} else {
    None
}

The only difference is the last line so that we still can enter the `GeneratorNestedColumnAliasing` branch if no rewrite happens?

Comment (minyyy) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:336]: The only logic change is that I changed
```val requiredAttrs = p.references -- g.producedAttributes ++ g.generator.references```
to 
```val requiredAttrs = p.references ++ g.generator.references```
The other changes is due to `prunedChild` is defined in the `ColumnPruning` class so I just inlined its logic.
For the change to `requiredAttrs`, the reason is that I think the previous logic is incorrect, I don't think `g.producedAttributes` should be removed from the `requiredAttrs` list.

Comment (sigmod) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:336]: Is it possible to not replicate the logic of `prunedChild`?  
`prunedChild` is used by every other branches. In case the logic in `prunedChild` evolve in the future, this call site will not get the benefit?

>> I don't think g.producedAttributes should be removed from the requiredAttrs list.

iiuc, `g.producedAttributes` shouldn't exist in `g.child.output`?  If so, why do we need to include them in `requiredAttrs`?

Comment (minyyy) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:334]: > Is it possible to not replicate the logic of `prunedChild`?

Done.

> iiuc, `g.producedAttributes` shouldn't exist in `g.child.output`? If so, why do we need to include them in `requiredAttrs` for `g.child`?

Sorry let me rephrase myself. The previous implementation is not incorrect, but it is redundant IMO. I think `g. producedAttributes` should not exist in `g.child.output`, even before this rule. But I am fine to add it back as I think it does no harm.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala (hunk 3 above)
Comment (minyyy) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala]: Done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala (hunk 4 above)
Comment (sigmod) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:334]: >> The previous implementation is not incorrect, but it is redundant IMO.

I think the meaning of `prunedChild` is a bit vague, particularly what `allReferences` mean. 
Therefore, "but it is redundant" is based on reading the content of `prunedChild`, rather than per the contract of `prunedChild`.

A cleaner version of it imo looks like the following:
```
def prunedChild(c: LogicalPlan, requiredAttributes: AttributeSet): LogicalPlan =
    require(requiredAttributes.subsetOf(c.outputSet))
    if (c.outputSet != requiredAttributes) {
      Project(requiredAttributes, c)
    } else {
      c
    }
```

I'm fine with go with your current code. We can update `prunedChild` in a followup PR -- maybe it'll fail some tests.

**Review by sigmod — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala (hunk 4 above)
Comment (sigmod) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:334]: Nit: it seems that we don't really need `var pruned`:

```
if (!newChild.fastEq(g.Child) || unrequiredIndices.toSet != g.unrequiredChildIndex.toSet) {
  p.copy(child = g.copy(child = newChild, unrequiredChildIndex = unrequiredIndices)) 
} else {
  None 
}
```

`||` can short circuit the second set equality check condition. But I think in reality, your code should also be fine. So it's your call.

**Comment by minyyy** [comment]
@cloud-fan Could you help reviewing this change please? Thanks.

**Review by sigmod — APPROVED** [review]
Thanks, @minyyy!

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasingSuite.scala [hunk 5]
```diff
@@ -313,6 +313,25 @@ class NestedColumnAliasingSuite extends SchemaPruningTest {
     comparePlans(optimized, expected)
   }
 
+  test("SPARK-38531: Nested field pruning for Project and PosExplode") {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasingSuite.scala:316]: I think the test in `ColumnPruningSuite` is sufficient. We didn't change `NestedColumnAliasing` object in this PR.

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 2 above)
Comment (minyyy) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: It is fixed in the latest implementation, I will mark it as resolved.

**Comment by cloud-fan** [comment]
thanks, merging to master/3.3!

**Comment by HyukjinKwon** [comment]
Actually let's revert this for now since this was more an improvement, and I think the workload it broke looks pretty basic ..

**Comment by HyukjinKwon** [comment]
Reverted at https://github.com/apache/spark/commit/161c596cafea9c235b5c918d8999c085401d73a9 and https://github.com/apache/spark/commit/4512e0943036d30587ab19a95efb0e66b47dd746, and created a PR to add a test case at https://github.com/apache/spark/pull/37084
